### PR TITLE
Optimize checking if the file is relevant for the cop

### PR DESCRIPTION
**Pull request by fatkodima**
Follow up to #11240.

Tested on Gitlab's codebase (`app/models` directory).

Because the project has a large set of rubocop todo files (https://gitlab.com/gitlab-org/gitlab/-/tree/master/.rubocop_todo), most items in `Exclude`/`Include` of which are exact file names (which I suppose correlates with most of the projects with todo files), it is wasteful to linearly check the list of patterns over and over to check if the file is relevant to the cop. 
Example of such configuration - https://gitlab.com/gitlab-org/gitlab/-/blob/master/.rubocop_todo/gitlab/namespaced_class.yml.

I optimized to firstly do cheaper checks first via partitioning an array of patterns into a set of exact match strings (and be able to do O(1) search) and the rest of the patterns.

## Before
Time: `86.4s`

```
==================================
  Mode: wall(1000)
  Samples: 86352 (0.09% miss rate)
  GC: 3948 (4.57%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     18789  (21.8%)       11671  (13.5%)     RuboCop::PathUtil#match_path?
      6742   (7.8%)        6742   (7.8%)     RuboCop::PathUtil#maybe_hidden_file?
      4925   (5.7%)        3005   (3.5%)     RuboCop::AST::Descendence#each_child_node
      2814   (3.3%)        2814   (3.3%)     (sweeping)
      2475   (2.9%)        2475   (2.9%)     block (2 levels) in <class:Node>
      3383   (3.9%)        2304   (2.7%)     Parser::Lexer#advance
      2089   (2.4%)        2076   (2.4%)     RuboCop::Config#for_cop
      1935   (2.2%)        1935   (2.2%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      2920   (3.4%)        1639   (1.9%)     RuboCop::Cop::Base#cop_config
      1388   (1.6%)        1388   (1.6%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      1210   (1.4%)        1210   (1.4%)     RuboCop::QAHelpers#in_qa_file?
      1203   (1.4%)        1203   (1.4%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      1128   (1.3%)        1128   (1.3%)     (marking)
      1081   (1.3%)        1081   (1.3%)     RuboCop::AST::Node#node_parts
      1067   (1.2%)        1067   (1.2%)     Parser::Source::Buffer#slice
      2802   (3.2%)         994   (1.2%)     RuboCop::Cop::Registry#enabled?
      1293   (1.5%)         964   (1.1%)     Parser::Source::Buffer#line_index_for_position
      1179   (1.4%)         899   (1.0%)     AST::Node#initialize
       757   (0.9%)         757   (0.9%)     Parser::Source::Range#initialize
```

## After
Time: `72.7s` (16% speedup) 🔥 

```
==================================
  Mode: wall(1000)
  Samples: 72528 (0.16% miss rate)
  GC: 4218 (5.82%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
      5123   (7.1%)        3154   (4.3%)     RuboCop::AST::Descendence#each_child_node
      2957   (4.1%)        2957   (4.1%)     (sweeping)
      2648   (3.7%)        2648   (3.7%)     block (2 levels) in <class:Node>
      3370   (4.6%)        2252   (3.1%)     Parser::Lexer#advance
      2234   (3.1%)        2234   (3.1%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      2202   (3.0%)        2182   (3.0%)     RuboCop::Config#for_cop
      1474   (2.0%)        1474   (2.0%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      2672   (3.7%)        1464   (2.0%)     RuboCop::Cop::Base#cop_config
      1432   (2.0%)        1432   (2.0%)     RuboCop::QAHelpers#in_qa_file?
      1256   (1.7%)        1256   (1.7%)     (marking)
      1247   (1.7%)        1247   (1.7%)     Gitlab::Styles::Rubocop::ModelHelpers#in_model?
      1499   (2.1%)        1247   (1.7%)     RuboCop::PathUtil.match_path?
      1139   (1.6%)        1139   (1.6%)     RuboCop::AST::Node#node_parts
      1125   (1.6%)        1125   (1.6%)     Parser::Source::Buffer#slice
      2987   (4.1%)        1038   (1.4%)     RuboCop::Cop::Registry#enabled?
      1367   (1.9%)        1017   (1.4%)     Parser::Source::Buffer#line_index_for_position
      1240   (1.7%)         943   (1.3%)     AST::Node#initialize
       833   (1.1%)         833   (1.1%)     Parser::Source::Range#initialize
       713   (1.0%)         713   (1.0%)     RuboCop::AST::ProcessedSource#valid_syntax?
       947   (1.3%)         705   (1.0%)     RuboCop::Cop::Base#initialize
       618   (0.9%)         617   (0.9%)     RuboCop::Cop::Base.badge
       606   (0.8%)         606   (0.8%)     RuboCop::AST::SendNode#first_argument_index
```

### CODE REVIEW

File: lib/rubocop/optimized_patterns.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module RuboCop
+  # @api private
+  module OptimizedPatterns
```
Comment (bbatsov) [inline lib/rubocop/optimized_patterns.rb:5]: I guess `OptimizedFilePatterns` (or just `FilePatterns`)  would be a bit more descriptive name.

File: lib/rubocop/optimized_patterns.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    # A wrapper around patterns array to perform optimized search.
```
Comment (bbatsov) [inline lib/rubocop/optimized_patterns.rb:6]: You can include a bit more of the reasoning for this class that you've mentioned in the PR description.

**Comment by bbatsov** [comment]
> Time: 72.7s (16% speedup) 🔥

That looks mighty impressive!

File: lib/rubocop/optimized_patterns.rb (hunk 1 above)
Comment (bbatsov) [reply lib/rubocop/optimized_patterns.rb:5]: Or even `PathPatterns`! Naming is hard! :D

**Comment by bbatsov** [comment]
I'll merge this now, as it's a private API and I'm just about to cut a new release. Please, tweak it when you have some time to do so.

**Comment by splattael** [comment]
I am very glad to see that some of the optimizations got into RuboCop :tada: Thanks a ton @fatkodima :bow: 

At GitLab we had ([well currently still have](https://gitlab.com/gitlab-org/gitlab/-/blob/1a7a13d270463bccc9ce37b16da4eac1e59ad166/rubocop/rubocop.rb#L6-7)) a set of patches to improve RuboCop's speed so now they can be removed :heart: 

See also https://gitlab.com/gitlab-org/gitlab/-/issues/377469

There's also https://github.com/rubocop/rubocop/compare/master...splattael:rubocop:speed-up which had a (very) similar approach to optimize path matching with additional support for Regexp patterns.

Again, thanks for your work @fatkodima :bow:

**Pull request merged**